[zprezto-update] Add convenience function to update zprezto

samcv opened the pull request:
This function checks if there is any update to zprezto, and if so
will pull in the changes. It will not attempt a pull unless
it is fastforwardable. It also makes sure the user is on the master
branch before attempting.

Yoshyn commented:
Pretty cool PR. 
Just one question : I have modification on theses files : 
  * runcoms/zpreztorc (adding module, & co ...)
  * runcoms/zshrc        (export EDITOR & VISUAL...

So, if i run this it fail due to : 
```
error: cannot pull with rebase: You have unstaged changes.
error: please commit or stash them.
```

Manually, i stash it and reapply after.
I think that a lot people modify theses files. Do you plan to manage it?

diraol commented:
Following up @Yoshyn comment, what could have been done here is something like what is done on `spf13-vim` distro.
We could have a `zprestorc.local` and a `zshrc.local` files that are imported on the end of the `zprestorc` and `zshrc` files. Doing so, the `.local` won't be versioned and then the users personalized configurations wouldn't get in conflict with the main config files, making this updates easier.
Does it make sense?

Also, what do you think about printing, as output from the update command, the messages from the commits that were pulled, so the user can know what was updated? Does it make sense, at least with an optional argument?

Yoshyn commented:
Well, initially i though about someting like that : 
```
printf "Try to stash before pulling..."
local zprezto_stash="zprezto-update-`date +%s`"
git stash save "$(echo $zprezto_stash)" --include-untracked
if git pull --ff-only; then
    git submodule update --recursive
    git stash pop
else
    printf "Unable to fast-forward the changes. You can fix this by running 'cd %s' and then 'git pull' to manually pull and possibly merge in changes\n"
fi
```

But it's just unmaintainable.  
After reading @diraol, i'm completely convinced. A local setup not versioned will be awesome

belak commented on init.zsh:
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+# zprezto convenience updater
+function zprezto-update {
+    local OWD="$(pwd)"
+    cd -- "${ZPREZTODIR}"
```
`pushd` and `popd` would probably be better for this. Then we don't need to track `OWD`. Also note that we should use `-q` to ensure the chpwd functions are not run.

samcv replied:
Fixed this in the latest commit.

belak commented on init.zsh:
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+                printf "Syncing submodules\n"
+                git submodule update --recursive
+            else
+                printf "Unable to fast-forward the changes. You can fix this by running 'cd %s' and then 'git pull' to manually pull and possibly merge in changes\n"
```
Looks like this is missing the param for the `%s`

samcv replied:
Fixed.

belak commented on init.zsh:
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+    local orig_branch="$(git symbolic-ref HEAD 2>/dev/null | cut -d '/' -f 3)"
+    if [[ "$orig_branch" == "master" ]]; then
+        git fetch
+        if [[ "$(git status -uno | grep behind)" ]]; then
```
This check seems a bit error prone. Maybe something like what's mentioned here would be better: https://stackoverflow.com/a/3278427 (on a side note, I had no idea `@` even existed in git)

samcv replied:
Took that suggestion and implemented it in my latest commit.

samcv commented:
Fixed @belak's suggestions in my latest commit and cleaned up a few things regarding the error handling. As stated in the commit description.

belak commented on init.zsh:
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+       printf "to manually pull and possibly merge in changes\n"
+   }
+   pushd -q
+   cd -- "${ZPREZTODIR}"
```
This would be better as `pushd -q -- "${ZPREZTODIR}"` so cd won't cause chpwd hooks to run (cd also has a similar option, but they can be combined, so it makes sense to)

belak commented on init.zsh:
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+function zprezto-update {
+   function cannot-fast-forward {
+       local STATUS="$1"
+       [[ "${STATUS}" ]] && printf "%s\n" "${STATUS}"
```
Should this be `[[ -n "${STATUS}" ]]`?

samcv replied:
`[[ -n "${STATUS}" ]]` is the same thing as `[[ "${STATUS}" ]]`.

Should I explicitly add the -n? (this works without -n even in posix shell by the way)

belak replied:
This is fine, I think... because we control the value of STATUS. If we didn't the -n would make sense.

belak commented on init.zsh:
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+       printf "zprezto install at '%s' is not on the master branch " "${ZPREZTODIR}"
+       printf "(you're on '%s')\nUnable to automatically update.\n" "${orig_branch}"
+   fi
+   popd -q
```
It would be nice to have proper return codes if this fails. `popd` could be moved to an atexit function (`trap 'popd -q' EXIT`) which would make it easier to return without having to copy the popd everywhere.

samcv replied:
Okay, I had to make a change and not run pushd -q earlier on to make popd -q work inside the trap. Otherwise it would take me back to the wrong folder.

belak commented on runcoms/zprofile:
```diff
@@ -65,12 +65,9 @@ fi
 # Temporary Files
 #
-#
-if [[ -z "$TMPDIR" ]]; then
```
This change seems unrelated...

belak commented on init.zsh:
```diff
@@ -17,6 +17,51 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+    printf "running\ncd '%s' and then\n'git pull' " "${ZPREZTODIR}"
+    printf "to manually pull and possibly merge in changes\n"
+  }
+  cd -- "${ZPREZTODIR}" || return 7
```
This will still trigger chpwd callbacks (which we don't want to run). There's an option to cd which won't trigger them... Also, I'm a bit confused why pushd didn't work...

samcv commented:
I've added the `-q` option to `cd`. Instead of using pushd and trying to popd on exit, I changed it to use `( )` around the function which keeps the contents in the function outside the external scope and avoids the issue of having to change back. Since the `cd` happens inside a different scope.

belak commented:
Alright, a subshell works. I'd like to see a comment mentioning that (because it's easy to miss that `()` was used rather than `{}`.

Once that's done and a small blurb is added to the readme (or somewhere relevant) I think this is ready.

samcv commented:
@belak I've added information to the README and added a comment to `init.zsh` about it being in a subshell.

belak commented on init.zsh:
```diff
@@ -17,6 +17,52 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
+  cd -q -- "${ZPREZTODIR}" || return 7
+  local orig_branch="$(git symbolic-ref HEAD 2>/dev/null | cut -d '/' -f 3)"
+  if [[ "$orig_branch" == "master" ]]; then
+    git fetch
```
Sorry, forgot about this one. We should probably return if git fetch fails.

samcv replied:
Done.

belak commented:
Merged! Thanks for your work on this!

samcv commented:
We can probably have an enhancement based on @Yoshyn's suggestion to possibly stash things and then reapply but I'd think it'd be best to keep it simple for at least a little while and make sure there's no issues with things as they have been merged in currently. Since at least we know it won't make any mistakes if it aborts in case of conflicts.

belak commented:
Yep! I figured, since this is new and not used by many people it's alright to have a few non-critical bugs... and this PR has had enough feedback on it. :)
